libcudacxx/include/cuda/std/__cccl/builtin.h

```diff
@@ -0,0 +1,429 @@
+#endif // __check_builtin(array_rank)
+
+#if __has_builtin(__array_extent)
+#  define _CCCL_BUILTIN_ARRAY_EXTENT(...) __array_extent(__VA_ARGS__)
+#endif // __has_builtin(array_extent)
+
+// nvhpc has a bug where it supports __builtin_addressof but does not mark it via __check_builtin
+#if __check_builtin(builtin_addressof) || (defined(_CCCL_COMPILER_GCC) && _CCCL_GCC_VERSION >= 70000) \
+  || defined(_CCCL_COMPILER_MSVC) || defined(_CCCL_COMPILER_NVHPC)
+#  define _CCCL_BUILTIN_ADDRESSOF(...) __builtin_addressof(__VA_ARGS__)
+#endif // __check_builtin(builtin_addressof)
+
+#if __check_builtin(builtin_assume) && !defined(_CCCL_CUDACC_BELOW_11_2)
+#  define _CCCL_BUILTIN_ASSUME(...) __builtin_assume(__VA_ARGS__)
+#endif // __check_builtin(builtin_assume) && nvcc >= 11.2
+
+// MSVC supports __builtin_bit_cast from 19.25 on
+// clang-9 supports __builtin_bit_cast but it is not a constant expression
+#if (__check_builtin(builtin_bit_cast) || (defined(_CCCL_COMPILER_MSVC) && _CCCL_MSVC_VERSION > 1925)) \
+  && !defined(_CCCL_CUDACC_BELOW_11_7) && !(defined(_CCCL_COMPILER_CLANG) && _CCCL_CLANG_VERSION < 100000)
+#  define _CCCL_BUILTIN_BIT_CAST(...) __builtin_bit_cast(__VA_ARGS__)
+#endif // __check_builtin(builtin_bit_cast)
+
+#if __check_builtin(builtin_contant_p) || defined(_CCCL_COMPILER_GCC)
+#  define _CCCL_BUILTIN_CONSTANT_P(...) __builtin_constant_p(__VA_ARGS__)
+#endif // __check_builtin(builtin_contant_p)
+
+#if __check_builtin(builtin_is_constant_evaluated) || (defined(_CCCL_COMPILER_GCC) && _CCCL_GCC_VERSION >= 90000) \
+  || (defined(_CCCL_COMPILER_MSVC) && _CCCL_MSVC_VERSION > 1924 && !defined(_CCCL_CUDACC_BELOW_11_3))
+#  define _CCCL_BUILTIN_IS_CONSTANT_EVALUATED(...) __builtin_is_constant_evaluated(__VA_ARGS__)
+#endif // __check_builtin(builtin_is_constant_evaluated)
+
+// NVCC and NVRTC in C++11 mode freaks out about `__builtin_is_constant_evaluated`.
+#if _CCCL_STD_VER < 2014 \
+  && (defined(_CCCL_CUDA_COMPILER_NVCC) || defined(_CCCL_COMPILER_NVRTC) || defined(_CCCL_COMPILER_NVHPC))
+#  undef _CCCL_BUILTIN_IS_CONSTANT_EVALUATED
+#endif // _CCCL_STD_VER < 2014 && _CCCL_CUDA_COMPILER_NVCC
+
+#if (__check_builtin(builtin_launder) || (defined(_CCCL_COMPILER_GCC) && _CCCL_GCC_VERSION >= 70000)) \
+  && (!defined(_CCCL_COMPILER_CLANG) || _CCCL_CLANG_VERSION >= 100000 || !defined(_CCCL_CUDACC_BELOW_11_3))
+#  define _CCCL_BUILTIN_LAUNDER(...) __builtin_launder(__VA_ARGS__)
+#endif // __check_builtin(builtin_launder) && gcc >= 7
+
+#if __check_builtin(__builtin_operator_new) && __check_builtin(__builtin_operator_delete) \
+  && defined(_CCCL_CUDA_COMPILER_CLANG)
+#  define _CCCL_BUILTIN_OPERATOR_DELETE(...) __builtin_operator_delete(__VA_ARGS__)
+#  define _CCCL_BUILTIN_OPERATOR_NEW(...)    __builtin_operator_new(__VA_ARGS__)
+#endif // __check_builtin(__builtin_operator_new) && __check_builtin(__builtin_operator_delete)
+
+#if __has_builtin(__decay) && defined(_CCCL_CUDA_COMPILER_CLANG)
```

Comment:
I wrote that at the top.

The `__check_builtin` facility is a clutch to work with older GCC / clang that do not support `__has_builtin`

However, that has some drawbacks, because the `__has_keyword` or `__has_feature` checks can lead to false positives.

So for all newer builtins that are only supported by new compilers anyhow, I am moving towards just using `__has_builtin`